File: wiz_dashboard/domain/reconcile.py
"""Pure cross-scan reconciliation — the heart of correct MTTR.

Given the current scan's findings and the prior ledger, this produces the updated ledger,
a per-scan ``observations`` log and the scan deltas. It has NO database or Streamlit
dependency so it is exhaustively unit-testable; ``data.ledger`` is the thin layer that
loads the prior ledger, calls ``reconcile`` and writes the result in one transaction.

Lifecycle rules:
  * First sighting      → OPEN, ``first_seen = min(API firstDetectedAt, scan ts)``.
  * Persisting (OPEN)   → advance ``last_seen``; keep ``first_seen`` earliest-known.
  * API-resolved        → ``resolvedAt`` present or status in ``RESOLVED_STATUSES``.
  * Disappearance       → was OPEN and present in the *immediately previous* scan but
                          absent now → resolved at the current scan ts (the usual
                          real-world remediation signal).
  * Reopen              → a RESOLVED vuln reappears as *active* (API not resolved) →
                          OPEN again, ``reopened_count++``, ``first_seen`` reset to start
                          a new episode.
"""

import json
from datetime import datetime, timezone

import pandas as pd

from wiz_dashboard.config import RESOLVED_STATUSES
from wiz_dashboard.domain.lifecycle import field, vuln_key
from wiz_dashboard.domain.severity import normalize_severity


def _clean(v):
    """Return ``None`` for missing/NaN/NaT/empty scalars, else the value unchanged."""
    if v is None:
        return None
    try:
        if not isinstance(v, (list, dict)) and pd.isna(v):
            return None
    except (TypeError, ValueError):
        pass
    if isinstance(v, str) and not v.strip():
        return None
    return v


def _parse(ts):
    """Parse an ISO timestamp (``Z`` or offset) to an aware UTC datetime, or ``None``."""
    ts = _clean(ts)
    if ts is None:
        return None
    if isinstance(ts, datetime):
        return ts if ts.tzinfo else ts.replace(tzinfo=timezone.utc)
    s = str(ts).strip()
    if s.endswith("Z"):
        s = s[:-1] + "+00:00"
    try:
        dt = datetime.fromisoformat(s)
    except ValueError:
        return None
    return dt if dt.tzinfo else dt.replace(tzinfo=timezone.utc)


def _iso(dt):
    """Canonical UTC ISO (``…Z``) for an aware datetime, or ``None``."""
    if dt is None:
        return None
    return dt.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def _min_iso(*values):
    """Earliest of the given timestamp values, as canonical ISO (``None`` if all empty)."""
    parsed = [d for d in (_parse(v) for v in values) if d is not None]
    return _iso(min(parsed)) if parsed else None


def _midpoint_iso(a, b):
    """Canonical ISO halfway between two timestamps (falls back to whichever parses)."""
    da, db = _parse(a), _parse(b)
    if da is None or db is None:
        return _iso(db) or _iso(da)
    return _iso(da + (db - da) / 2)


_TAGS_PREFIX = "vulnerableAsset.tags."


def _tags_json(record):
    """The asset's tags as canonical JSON (sorted keys), or ``None`` when absent.

    Accepts the nested raw node (``vulnerableAsset.tags`` dict) and the flattened
    ``vulnerableAsset.tags.<key>`` record shape. Sorted keys keep delete→rebuild and
    compaction-checkpoint replays byte-stable.
    """
    va = record.get("vulnerableAsset")
    tags = va.get("tags") if isinstance(va, dict) else None
    if not isinstance(tags, dict):
        flat = record.get("vulnerableAsset.tags")
        tags = flat if isinstance(flat, dict) else None
    if tags is None:
        tags = {
            k[len(_TAGS_PREFIX):]: v
            for k, v in record.items()
            if k.startswith(_TAGS_PREFIX) and _clean(v) is not None
        }
    tags = {str(k): v for k, v in tags.items() if _clean(v) is not None or v == ""}
    if not tags:
        return None
    try:
        return json.dumps(tags, sort_keys=True, ensure_ascii=False)
    except (TypeError, ValueError):
        return None


def _make_row(record, key, sev, first_seen, scan_id, scan_ts):
    return {
        "vuln_key": key,
        "cve": _clean(record.get("name")),
        "severity": sev,
        "asset_id": field(record, "vulnerableAsset.id") or None,
        "asset_name": field(record, "vulnerableAsset.name") or None,
        "asset_type": field(record, "vulnerableAsset.type") or None,
        "cloud": field(record, "vulnerableAsset.cloudPlatform") or None,
        "subscription_name": field(record, "vulnerableAsset.subscriptionName") or None,
        "subscription_ext_id": field(
            record, "vulnerableAsset.subscriptionExternalId", "vulnerableAsset.subscriptionId"
        )
        or None,
        "tags_json": _tags_json(record),
        "first_seen": first_seen,
        "last_seen": scan_ts,
        "status": "OPEN",
        "resolved_at": None,
        "resolution_src": None,
        "reopened_count": 0,
        "first_scan_id": scan_id,
        "last_scan_id": scan_id,
    }


def reconcile(
    current_records,
    existing_ledger,
    scan_id,
    scan_ts,
    prev_scan_id,
    *,
    disappearance_mode="scan_ts",
    prev_scan_ts=None,
    scanned_severities=None,
    prev_scan_id_by_severity=None,
):
    """Reconcile one flat scan against the prior ledger.

    Args:
        current_records: list of dict, one per finding (flat shape; dotted or nested keys).
        existing_ledger: ``{vuln_key: row}`` loaded from the durable base.
        scan_id / scan_ts: identity + timestamp of this scan (ISO; usually equal).
        prev_scan_id: the immediately-previous scan's id, or ``None`` for the first scan.
        disappearance_mode: ``"scan_ts"`` (default) or ``"midpoint"`` for the inferred
            resolution timestamp; ``prev_scan_ts`` is required for ``"midpoint"``.
        scanned_severities: the severity scope of THIS scan (iterable of normalized
            uppercase values), or ``None`` for an unscoped scan. Out-of-scope OPEN rows
            are exempt from disappearance resolution — their absence is expected, not a
            remediation signal. Their lifecycle simply pauses until they're scanned again.
        prev_scan_id_by_severity: ``{severity: scan_id}`` of the most recent prior scan
            whose scope *included* each severity. Replaces ``prev_scan_id`` in the
            disappearance guard per-severity, so a finding that vanished while its
            severity went unscanned still resolves on the first scan that covers it
            again (``resolved_at`` = that scan's ts — conservative, same tradeoff as
            ``DISAPPEARANCE_RESOLUTION="scan_ts"``). Severities missing from the map
            fall back to ``prev_scan_id``.

    Returns:
        ``(updated_ledger, observations, deltas)`` where ``deltas`` is
        ``{new_count, resolved_count, reopened_count}``.
    """
    # Ledger rows are flat dicts of scalars, so a per-row shallow copy gives the same
    # don't-mutate-the-input guarantee as deepcopy at a fraction of the cost (deepcopy
    # walked every cell of a 100k+-row ledger on every scan).
    updated = {key: dict(row) for key, row in existing_ledger.items()}
    seen = set()
    observations = []
    new_count = resolved_count = reopened_count = 0

    scan_ts_iso = _iso(_parse(scan_ts)) or str(scan_ts)

    for rec in current_records:
        key = vuln_key(rec)
        if key in seen:  # duplicate finding within the same scan — first wins
            continue
        seen.add(key)

        sev = normalize_severity(_clean(rec.get("severity")))
        api_first = (
            _clean(rec.get("firstDetectedAt"))
            or _clean(rec.get("firstSeenAt"))
            or _clean(rec.get("createdAt"))
        )
        api_status = str(_clean(rec.get("status")) or "").upper()
        api_resolved = (
            _clean(rec.get("resolvedAt"))
            or _clean(rec.get("remediatedAt"))
            or _clean(rec.get("fixedAt"))
        )
        api_says_resolved = bool(api_resolved) or api_status in RESOLVED_STATUSES

        row = updated.get(key)
        if row is None:
            first_seen = _min_iso(api_first, scan_ts_iso) or scan_ts_iso
            row = _make_row(rec, key, sev, first_seen, scan_id, scan_ts_iso)
            updated[key] = row
            new_count += 1
        elif row["status"] == "RESOLVED" and not api_says_resolved:
            # Genuine reopen: a previously-resolved vuln is active again. Start a new
            # episode so the next resolution measures THIS episode, not the original.
            row["status"] = "OPEN"
            row["resolved_at"] = None
            row["resolution_src"] = None
            row["reopened_count"] = int(row.get("reopened_count", 0)) + 1
            row["first_seen"] = _min_iso(api_first, scan_ts_iso) or scan_ts_iso
            row["last_seen"] = scan_ts_iso
            row["last_scan_id"] = scan_id
            reopened_count += 1
        else:
            # Persisting (OPEN) or a still-resolved finding being re-listed. Keep
            # first_seen earliest-known; never let it drift later.
            if row["status"] == "OPEN":
                row["first_seen"] = (
                    _min_iso(row.get("first_seen"), api_first) or row.get("first_seen")
                )
            row["last_seen"] = scan_ts_iso
            row["last_scan_id"] = scan_id

        # Latest observation wins for display attributes.
        row["severity"] = sev
        row["cve"] = _clean(rec.get("name"))
        row["asset_id"] = field(rec, "vulnerableAsset.id") or row.get("asset_id")
        row["asset_name"] = field(rec, "vulnerableAsset.name") or row.get("asset_name")
        row["asset_type"] = field(rec, "vulnerableAsset.type") or row.get("asset_type")
        row["cloud"] = field(rec, "vulnerableAsset.cloudPlatform") or row.get("cloud")
        row["subscription_name"] = field(rec, "vulnerableAsset.subscriptionName") or row.get(
            "subscription_name"
        )
        row["subscription_ext_id"] = field(
            rec, "vulnerableAsset.subscriptionExternalId", "vulnerableAsset.subscriptionId"
        ) or row.get("subscription_ext_id")
        row["tags_json"] = _tags_json(rec) or row.get("tags_json")

        # API-declared resolution closes a currently-open row.
        if api_says_resolved and row["status"] == "OPEN":
            row["status"] = "RESOLVED"
            row["resolved_at"] = _iso(_parse(api_resolved)) if api_resolved else scan_ts_iso
            row["resolution_src"] = "api"
            resolved_count += 1

        observations.append(
            {
                "scan_id": scan_id,
                "vuln_key": key,
                "present": 1,
                "severity": sev,
                "status": row["status"],
            }
        )

    # Disappearance: OPEN vulns present in the immediately-previous scan but absent now.
    if prev_scan_id is not None:
        scope = set(scanned_severities) if scanned_severities is not None else None
        for key, row in updated.items():
            if key in seen or row["status"] == "RESOLVED":
                continue
            sev_row = row.get("severity")
            if scope is not None and sev_row not in scope:
                # This severity wasn't scanned — absence is expected, not resolution.
                continue
            expected_prev = (prev_scan_id_by_severity or {}).get(sev_row, prev_scan_id)
            if row.get("last_scan_id") != expected_prev:
                continue
            if disappearance_mode == "midpoint" and prev_scan_ts:
                row["resolved_at"] = _midpoint_iso(prev_scan_ts, scan_ts_iso)
            else:
                row["resolved_at"] = scan_ts_iso
            row["status"] = "RESOLVED"
            row["resolution_src"] = "disappeared"
            resolved_count += 1
            observations.append(
                {
                    "scan_id": scan_id,
                    "vuln_key": key,
                    "present": 0,
                    "severity": row.get("severity"),
                    "status": "RESOLVED",
                }
            )

    deltas = {
        "new_count": new_count,
        "resolved_count": resolved_count,
        "reopened_count": reopened_count,
    }
    return updated, observations, deltas

File: gas/src/domain/reconcile.ts
// Pure cross-scan reconciliation — the port of wiz_dashboard/domain/reconcile.py.
//
// Lifecycle rules:
//   * First sighting      -> OPEN, first_seen = min(API firstDetectedAt, scan ts).
//   * Persisting (OPEN)   -> advance last_seen; keep first_seen earliest-known.
//   * API-resolved        -> resolvedAt present or status in RESOLVED_STATUSES.
//   * Disappearance       -> was OPEN and present in the immediately previous scan but
//                            absent now -> resolved at the current scan ts.
//   * Reopen              -> a RESOLVED vuln reappears as active -> OPEN again,
//                            reopened_count++, first_seen reset (new episode).

import { RESOLVED_STATUSES } from "./config";
import { field, vulnKey } from "./lifecycle";
import { normalizeSeverity } from "./severity";
import { clean, midpointIso, minIso, parseTs, present, toIso, type Rec } from "./util";

export type LedgerRow = {
  vuln_key: string;
  cve: string | null;
  severity: string | null;
  asset_id: string | null;
  asset_name: string | null;
  asset_type: string | null;
  cloud: string | null;
  first_seen: string | null;
  last_seen: string | null;
  status: string;
  resolved_at: string | null;
  resolution_src: string | null;
  reopened_count: number;
  first_scan_id: string | null;
  last_scan_id: string | null;
  subscription_name: string | null;
  subscription_ext_id: string | null;
  tags_json: string | null;
  // Vendor-fix capture (actionable clock). fix_date: the upstream fix date the API
  // reports; fix_observed_at: the scan ts at which a fix signal (fixedVersion or
  // fixDate) was first seen for this episode. Both sticky first-wins; see reconcile().
  fix_date: string | null;
  fix_observed_at: string | null;
  // Exploit-intelligence capture (remediation coverage / efficiency). The Wiz node carries
  // these, but only the *current scan frame* does — a finding resolved by disappearance is
  // gone from the frame entirely, so measuring "was what we remediated actually high-risk?"
  // is impossible unless the signal is durable. Hence these columns. Sticky-MONOTONE, not
  // latest-wins like severity/cve: see seedRisk() in reconcile().
  //
  // null means NOT CAPTURED, and is emphatically not the same as false — a row whose signal
  // was never observed is UNCLASSIFIED and leaves both sides of every rate (see
  // domain/program.ts). Never default these to false.
  has_kev: boolean | null; // ever observed on the CISA KEV catalog (hasCisaKevExploit)
  has_exploit: boolean | null; // ever observed with a known exploit (hasExploit)
  epss: number | null; // PEAK observed epssProbability
  risk_observed_at: string | null; // scan ts at which any risk signal first arrived
};

export const LEDGER_COLUMNS: (keyof LedgerRow)[] = [
  "vuln_key", "cve", "severity", "asset_id", "asset_name", "asset_type", "cloud",
  "first_seen", "last_seen", "status", "resolved_at", "resolution_src",
  "reopened_count", "first_scan_id", "last_scan_id",
  "subscription_name", "subscription_ext_id", "tags_json",
  "fix_date", "fix_observed_at",
  "has_kev", "has_exploit", "epss", "risk_observed_at",
];

export interface Observation {
  scan_id: string;
  vuln_key: string;
  present: 0 | 1;
  severity: string | null;
  status: string;
}

export interface Deltas {
  new_count: number;
  resolved_count: number;
  reopened_count: number;
}

const TAGS_PREFIX = "vulnerableAsset.tags.";

/**
 * The asset's tags as canonical JSON (sorted keys), or null when absent. Accepts the
 * nested raw node (vulnerableAsset.tags dict) and flattened vulnerableAsset.tags.<key>
 * record shapes. Sorted keys keep delete-rebuild and checkpoint replays byte-stable.
 */
export function tagsJson(record: Rec): string | null {
  const va = record["vulnerableAsset"];
  let tags: Rec | null = null;
  if (va && typeof va === "object" && !Array.isArray(va)) {
    const t = (va as Rec)["tags"];
    if (t && typeof t === "object" && !Array.isArray(t)) tags = t as Rec;
  }
  if (tags === null) {
    const flat = record["vulnerableAsset.tags"];
    if (flat && typeof flat === "object" && !Array.isArray(flat)) tags = flat as Rec;
  }
  if (tags === null) {
    const collected: Rec = {};
    for (const [k, v] of Object.entries(record)) {
      if (k.startsWith(TAGS_PREFIX) && clean(v) !== null) {
        collected[k.slice(TAGS_PREFIX.length)] = v;
      }
    }
    tags = collected;
  }
  const kept: Rec = {};
  for (const [k, v] of Object.entries(tags)) {
    if (clean(v) !== null || v === "") kept[String(k)] = v;
  }
  const keys = Object.keys(kept).sort();
  if (!keys.length) return null;
  // Canonical JSON with sorted keys (json.dumps(sort_keys=True) parity for the flat
  // string/number/bool values tags actually carry).
  const parts = keys.map((k) => `${JSON.stringify(k)}: ${JSON.stringify(kept[k])}`);
  return `{${parts.join(", ")}}`;
}

function makeRow(
  record: Rec,
  key: string,
  sev: string,
  firstSeen: string | null,
  scanId: string,
  scanTs: string,
  fixDate: string | null,
  fixObservedAt: string | null,
): LedgerRow {
  return {
    vuln_key: key,
    cve: (clean(record["name"]) as string | null) ?? null,
    severity: sev,
    asset_id: field(record, "vulnerableAsset.id") || null,
    asset_name: field(record, "vulnerableAsset.name") || null,
    asset_type: field(record, "vulnerableAsset.type") || null,
    cloud: field(record, "vulnerableAsset.cloudPlatform") || null,
    subscription_name: field(record, "vulnerableAsset.subscriptionName") || null,
    subscription_ext_id:
      field(record, "vulnerableAsset.subscriptionExternalId", "vulnerableAsset.subscriptionId") ||
      null,
    tags_json: tagsJson(record),
    first_seen: firstSeen,
    last_seen: scanTs,
    status: "OPEN",
    resolved_at: null,
    resolution_src: null,
    reopened_count: 0,
    first_scan_id: scanId,
    last_scan_id: scanId,
    fix_date: fixDate,
    fix_observed_at: fixObservedAt,
    // Left empty here and filled by mergeRiskSignals() after the branch, which runs for new,
    // reopened, and persisting rows alike (the merge is identical in all three).
    ...emptyRiskSignals(),
  };
}

/** The mutable slice of a row the risk merge touches — ledger rows and episodes alike. */
export type RiskSignalFields = Pick<
  LedgerRow,
  "has_kev" | "has_exploit" | "epss" | "risk_observed_at"
>;

export function emptyRiskSignals(): RiskSignalFields {
  return { has_kev: null, has_exploit: null, epss: null, risk_observed_at: null };
}

/**
 * Hydrate the four risk columns off a stored row — a sheet row, a Drive snapshot row, or a
 * migration-bundle row. Blank/absent stays `null` (never coerced to `false` or `0`, which
 * would turn "not captured" into "captured, and the answer was no"); the plain-text ledger
 * grid round-trips booleans as the strings "TRUE"/"FALSE", which `observeRiskSignals`'
 * coercion already understands.
 */
export function coerceRiskSignals(r: Rec): RiskSignalFields {
  const obs = observeRiskSignals({
    hasCisaKevExploit: r["has_kev"],
    hasExploit: r["has_exploit"],
    epssProbability: r["epss"],
  });
  return {
    has_kev: obs.kev,
    has_exploit: obs.exploit,
    epss: obs.epss,
    risk_observed_at: (clean(r["risk_observed_at"]) as string | null) ?? null,
  };
}

/**
 * One observation's exploit-intelligence signals, each **null when the record does not
 * carry it at all** — which is emphatically not the same as an observed `false`. A slim
 * record written before these fields were captured simply lacks the keys; a tenant whose
 * Wiz plan omits EPSS returns null. Both must stay distinguishable from "observed, and the
 * answer was no", or coverage/efficiency silently impute a negative (see domain/program.ts).
 */
export function observeRiskSignals(
  rec: Rec,
): { kev: boolean | null; exploit: boolean | null; epss: number | null } {
  const bool = (v: unknown): boolean | null => {
    if (typeof v === "boolean") return v;
    // Sheets round-trip: the ledger grid is plain-text formatted (sheetsDb.ensureTabs), so a
    // boolean written with setValues reads back as the string "TRUE"/"FALSE".
    if (typeof v === "string") {
      const s = v.trim().toUpperCase();
      if (s === "TRUE") return true;
      if (s === "FALSE") return false;
    }
    return null;
  };
  const rawEpss = clean(rec["epssProbability"]);
  const n = typeof rawEpss === "number" ? rawEpss : rawEpss === null ? NaN : Number(rawEpss);
  return {
    kev: bool(rec["hasCisaKevExploit"]),
    exploit: bool(rec["hasExploit"]),
    epss: Number.isFinite(n) ? (n as number) : null,
  };
}

/**
 * Merge one observation's risk signals into a row, in place — **monotone, idempotent, and
 * order-independent**. Booleans go null → false → true and never back; `epss` keeps the peak
 * observed; `risk_observed_at` keeps the earliest witnessing scan.
 *
 * Deliberately NOT the latest-observation-wins treatment `severity` / `cve` / asset fields
 * get in reconcile(). Three reasons, all load-bearing:
 *
 *   * Exploit knowledge is monotone in reality — a CVE does not become un-exploited, and
 *     CISA KEV entries are effectively never withdrawn. EPSS genuinely decays, so *peak*
 *     EPSS is a deliberate choice: the question coverage asks is "was this something you
 *     should have prioritized", not "is it still scary today".
 *   * It keeps the high-risk label monotone, so a finding cannot silently leave the coverage
 *     denominator between scans. Without that, trend.withCoverageEfficiency would rewrite
 *     its own history on every scan — a point plotted last week would change value this
 *     week for reasons unrelated to any remediation work, which makes the series
 *     uninterpretable and the whole page unauditable.
 *   * A resolved row therefore freezes at its peak known risk, the conservative reading:
 *     coverage never under-counts what we ought to have fixed.
 *
 * Order-independence is what lets the archive backfill (server/backfillJobs.ts) replay saved
 * scans newest-first, resume after a crash, and be re-run from scratch, all converging on
 * byte-identical state without any bookkeeping.
 *
 * Note the divergence from `seedFix`: risk signals do **not** reset on reopen (the reopen
 * branch in reconcile() clears the vendor-fix clock, which is per-episode). Exploit
 * availability is a property of the vulnerability, not of the episode.
 */
export function mergeRiskSignals(row: RiskSignalFields, rec: Rec, scanTsIso: string): void {
  const obs = observeRiskSignals(rec);
  // `== null` (not `=== null`) also catches undefined on rows read back from a sheet
  // written before these columns existed.
  if (obs.kev !== null && (row.has_kev == null || obs.kev)) row.has_kev = obs.kev;
  if (obs.exploit !== null && (row.has_exploit == null || obs.exploit)) {
    row.has_exploit = obs.exploit;
  }
  if (obs.epss !== null && (row.epss == null || obs.epss > row.epss)) row.epss = obs.epss;
  const witnessed = obs.kev !== null || obs.exploit !== null || obs.epss !== null;
  if (!witnessed) return;
  // Earliest-wins, and genuinely a min rather than a first-wins guard: the backfill replays
  // scans newest-first, so a later call can legitimately carry an earlier timestamp.
  if (row.risk_observed_at == null || scanTsIso < row.risk_observed_at) {
    row.risk_observed_at = scanTsIso;
  }
}

export interface ReconcileOptions {
  disappearanceMode?: "scan_ts" | "midpoint";
  prevScanTs?: string | null;
  scannedSeverities?: string[] | null;
  prevScanIdBySeverity?: Record<string, string> | null;
}

/**
 * Reconcile one flat scan against the prior ledger.
 * Returns {ledger, observations, deltas}; neither input is mutated.
 */
export function reconcile(
  currentRecords: Rec[],
  existingLedger: Record<string, LedgerRow>,
  scanId: string,
  scanTs: string,
  prevScanId: string | null,
  options: ReconcileOptions = {},
): { ledger: Record<string, LedgerRow>; observations: Observation[]; deltas: Deltas } {
  const {
    disappearanceMode = "scan_ts",
    prevScanTs = null,
    scannedSeverities = null,
    prevScanIdBySeverity = null,
  } = options;

  // Rows are flat scalar dicts, so a shallow per-row copy preserves the inputs.
  const updated: Record<string, LedgerRow> = {};
  for (const [key, row] of Object.entries(existingLedger)) updated[key] = { ...row };

  const seen = new Set<string>();
  const observations: Observation[] = [];
  let newCount = 0;
  let resolvedCount = 0;
  let reopenedCount = 0;

  const scanTsIso = toIso(parseTs(scanTs)) ?? String(scanTs);

  for (const rec of currentRecords) {
    const key = vulnKey(rec);
    if (seen.has(key)) continue; // duplicate within the same scan — first wins
    seen.add(key);

    const sev = normalizeSeverity(clean(rec["severity"]));
    const apiFirst =
      clean(rec["firstDetectedAt"]) ?? clean(rec["firstSeenAt"]) ?? clean(rec["createdAt"]);
    const apiStatus = String(clean(rec["status"]) ?? "").toUpperCase();
    const apiResolved =
      clean(rec["resolvedAt"]) ?? clean(rec["remediatedAt"]) ?? clean(rec["fixedAt"]);
    const apiSaysResolved = present(apiResolved) || RESOLVED_STATUSES.has(apiStatus);

    // Vendor-fix signal for this observation: a concrete fixedVersion or a fixDate.
    // recFixDate is the normalized upstream fix date (null when unparseable/absent).
    const fixSignal = present(rec["fixedVersion"]) || present(rec["fixDate"]);
    const recFixDate = present(rec["fixDate"]) ? toIso(parseTs(rec["fixDate"])) : null;
    // Sticky first-wins: only ever fill a currently-empty field; never clear or overwrite.
    // `== null` also catches undefined from pre-migration snapshot rows lacking the columns.
    const seedFix = (r: LedgerRow): void => {
      if (r.fix_date == null && recFixDate !== null) r.fix_date = recFixDate;
      if (r.fix_observed_at == null && fixSignal) r.fix_observed_at = scanTsIso;
    };

    let row = updated[key];
    if (row === undefined) {
      const firstSeen = minIso(apiFirst, scanTsIso) ?? scanTsIso;
      row = makeRow(rec, key, sev, firstSeen, scanId, scanTsIso, recFixDate, fixSignal ? scanTsIso : null);
      updated[key] = row;
      newCount += 1;
    } else if (row.status === "RESOLVED" && !apiSaysResolved) {
      // Genuine reopen: start a new episode so the next resolution measures THIS
      // episode, not the original. The fix clock resets too — the prior episode's
      // fix is irrelevant — then re-seeds from the reopening record.
      row.status = "OPEN";
      row.resolved_at = null;
      row.resolution_src = null;
      row.reopened_count = Number(row.reopened_count ?? 0) + 1;
      row.first_seen = minIso(apiFirst, scanTsIso) ?? scanTsIso;
      row.last_seen = scanTsIso;
      row.last_scan_id = scanId;
      row.fix_date = null;
      row.fix_observed_at = null;
      seedFix(row);
      reopenedCount += 1;
    } else {
      // Persisting (OPEN) or a still-resolved finding being re-listed. Keep
      // first_seen earliest-known; never let it drift later.
      if (row.status === "OPEN") {
        row.first_seen = minIso(row.first_seen, apiFirst) ?? row.first_seen;
      }
      row.last_seen = scanTsIso;
      row.last_scan_id = scanId;
      seedFix(row);
    }

    // Exploit intelligence — one call for all three branches above, because the merge is
    // monotone and idempotent (unlike seedFix, which has to run per-branch around the reopen
    // reset). See mergeRiskSignals for why this is sticky rather than latest-wins.
    mergeRiskSignals(row, rec, scanTsIso);

    // Latest observation wins for display attributes.
    row.severity = sev;
    row.cve = (clean(rec["name"]) as string | null) ?? null;
    row.asset_id = field(rec, "vulnerableAsset.id") || row.asset_id;
    row.asset_name = field(rec, "vulnerableAsset.name") || row.asset_name;
    row.asset_type = field(rec, "vulnerableAsset.type") || row.asset_type;
    row.cloud = field(rec, "vulnerableAsset.cloudPlatform") || row.cloud;
    row.subscription_name =
      field(rec, "vulnerableAsset.subscriptionName") || row.subscription_name;
    row.subscription_ext_id =
      field(rec, "vulnerableAsset.subscriptionExternalId", "vulnerableAsset.subscriptionId") ||
      row.subscription_ext_id;
    row.tags_json = tagsJson(rec) ?? row.tags_json;

    // API-declared resolution closes a currently-open row.
    if (apiSaysResolved && row.status === "OPEN") {
      row.status = "RESOLVED";
      row.resolved_at = present(apiResolved) ? toIso(parseTs(apiResolved)) : scanTsIso;
      row.resolution_src = "api";
      resolvedCount += 1;
    }

    observations.push({
      scan_id: scanId,
      vuln_key: key,
      present: 1,
      severity: sev,
      status: row.status,
    });
  }

  // Disappearance: OPEN vulns present in the immediately-previous scan but absent now.
  if (prevScanId !== null) {
    const scope = scannedSeverities !== null ? new Set(scannedSeverities) : null;
    for (const [key, row] of Object.entries(updated)) {
      if (seen.has(key) || row.status === "RESOLVED") continue;
      const sevRow = row.severity;
      if (scope !== null && (sevRow === null || !scope.has(sevRow))) {
        // This severity wasn't scanned — absence is expected, not resolution.
        continue;
      }
      const expectedPrev =
        (prevScanIdBySeverity ?? {})[sevRow ?? ""] ?? prevScanId;
      if (row.last_scan_id !== expectedPrev) continue;
      if (disappearanceMode === "midpoint" && prevScanTs) {
        row.resolved_at = midpointIso(prevScanTs, scanTsIso);
      } else {
        row.resolved_at = scanTsIso;
      }
      row.status = "RESOLVED";
      row.resolution_src = "disappeared";
      resolvedCount += 1;
      observations.push({
        scan_id: scanId,
        vuln_key: key,
        present: 0,
        severity: row.severity,
        status: "RESOLVED",
      });
    }
  }

  return {
    ledger: updated,
    observations,
    deltas: {
      new_count: newCount,
      resolved_count: resolvedCount,
      reopened_count: reopenedCount,
    },
  };
}
